src/stash/src/persist.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::{hash_map, HashMap};
+use std::fmt;
+use std::hash::Hash;
+use std::vec;
+
+use ore::metrics::MetricsRegistry;
+use persist::client::{RuntimeClient, StreamWriteHandle};
+use persist::indexed::Snapshot;
+use persist::runtime::{self, RuntimeConfig};
+use persist::storage::{Blob, Log};
+use persist_types::Codec;
+use timely::progress::Antichain;
+
+use crate::{Stash, StashError, StashOp};
+
+type Error = StashError<persist::error::Error>;
+
+/// A stash backed by a [`persist`] collection.
+#[derive(Debug)]
+pub struct PersistStash<K, V> {
+    _client: RuntimeClient,
+    write_handle: StreamWriteHandle<K, V>,
+    data: HashMap<K, V>,
+    timestamp: u64,
+}
+
+impl<K, V> PersistStash<K, V>
+where
+    K: Codec + Clone + Eq + Hash + Ord + fmt::Debug,
+    V: Codec + Clone + Eq + Hash + Ord + fmt::Debug,
+{
+    /// Opens the stash specified by the provided blob and log.
+    pub fn open<B, L>(
+        blob: B,
+        log: L,
+        metrics_registry: &MetricsRegistry,
+    ) -> Result<PersistStash<K, V>, Error>
+    where
+        B: Blob + Send + 'static,
+        L: Log + Send + 'static,
+    {
+        let client = runtime::start(
+            RuntimeConfig::default(),
+            log,
+            blob,
+            build_info::DUMMY_BUILD_INFO,
+            metrics_registry,
+            None,
+        )?;
+        let (write_handle, read_handle) = client.create_or_load("stash");
+        write_handle.allow_compaction(Antichain::new());
+
+        let mut stage = vec![];
+        for entry in read_handle.snapshot()?.into_iter() {
+            let ((key, val), _ts, diff) = entry?;
+            stage.push(((key, val), diff));
+        }
+        differential_dataflow::consolidation::consolidate(&mut stage);
+
+        let mut data = HashMap::new();
+        for ((key, val), diff) in stage {
+            if diff != 1 {
+                return Err(StashError::Corruption(format!(
+                    "unexpected diff {} for key {:?}",
+                    diff, key
+                )));
+            }
+            data.insert(key, val);
+        }
+
```

Comment:
given that the semantics of this impl are to always write and seal data, it might also be worth being defensive here about the seal timestamp of the snapshot being greater than all of the data in it